src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/germline/IntegerCopyNumberExpectationsCalculator.java

```diff
@@ -0,0 +1,342 @@
+            }
+        }
+
+        /* calculate copy ratio posterior mean and variance */
+        final double[] logCopyRatioPriorMeans = hiddenStatePriorProbabilities.stream()
+                .mapToDouble(pdf -> calculateMeanDiscreteStates(includedHiddenStatesIndices, pdf,
+                        hiddenStatesLogCopyRatios))
+                .toArray();
+        final double[] logCopyRatioPriorVariances = IntStream.range(0, targetsList.size())
+                .mapToDouble(ti -> calculateMeanDiscreteStates(includedHiddenStatesIndices,
+                        hiddenStatePriorProbabilities.get(ti), hiddenStatesLogCopyRatiosSquared)
+                        - FastMath.pow(logCopyRatioPriorMeans[ti], 2))
+                .toArray();
+
+        if (CHECK_FOR_NANS) {
+            final int[] badTargets = IntStream.range(0, targetsList.size())
+                    .filter(ti -> Double.isNaN(logCopyRatioPriorMeans[ti]) ||
+                            Double.isNaN(logCopyRatioPriorVariances[ti])).toArray();
+            if (badTargets.length > 0) {
+                throw new RuntimeException("Some of the copy ratio prior expectations are ill-defined; targets: " +
+                        Arrays.stream(badTargets).mapToObj(String::valueOf).collect(Collectors.joining(", ", "[", "]")));
+            }
+        }
+
+        return new CopyRatioExpectations(logCopyRatioPriorMeans, logCopyRatioPriorVariances);
+    }
+
+    @Override
+    public CopyRatioHiddenMarkovModelResults<CoverageModelCopyRatioEmissionData,
+            IntegerCopyNumberState> getCopyRatioHiddenMarkovModelResults(@Nonnull final CopyRatioCallingMetadata copyRatioCallingMetadata,
+                                                                         @Nonnull final List<Target> activeTargets,
+                                                                         @Nonnull final List<CoverageModelCopyRatioEmissionData> emissionData) {
+        verifyArgsForHiddenMarkovModelResults(copyRatioCallingMetadata, activeTargets, emissionData);
+
+        /* choose the HMM */
+        final IntegerCopyNumberHiddenMarkovModel<CoverageModelCopyRatioEmissionData> genotypeSpecificHMM =
+                hmm.get(copyRatioCallingMetadata.getSampleSexGenotypeData().getSexGenotype());
+
+        /* supplement the emission data with sample metadata */
+        emissionData.forEach(datum -> datum.setCopyRatioCallingMetadata(copyRatioCallingMetadata));
+
+        /* run the forward-backward algorithm */
+        final ForwardBackwardAlgorithm.Result<CoverageModelCopyRatioEmissionData, Target, IntegerCopyNumberState> fbResult =
+                ForwardBackwardAlgorithm.apply(emissionData, activeTargets, genotypeSpecificHMM);
+
+        /* run the Viterbi algorithm */
+        final List<IntegerCopyNumberState> viterbiResult = ViterbiAlgorithm.apply(emissionData, activeTargets,
+                genotypeSpecificHMM);
+
+        return new CopyRatioHiddenMarkovModelResults<>(fbResult, viterbiResult);
```

Comment:
@samuelklee and I had a lengthy discussion about this. The Viterbi result, and what you get from stacking MAP on each target can be (very) different. One can think of the former as the ground state of the entire system, and the latter as the most favorable local state after tracing out the rest of the system.

What we thought would be interesting to do is to:
- segment based on Viterbi,
- segment based on stacking local CR MAP, and
- generate a swarm of hidden state samples from the HMM, segment each sample, and create a 1D density plot for state transition "hotspots".

The hope is that (1) gives us the best overall hidden chain (which may sacrifice local calls), (2) gives us target-resolved genotypes, and (3) gives us some insight about "excitations" about the Viterbi state, i.e. (3) is something between (1) and (2).

What do you think @davidbenjamin @samuelklee? while this in-depth output is intriguing for us to look at and study, the typical user just wants one list of "CNV calls".